### PR TITLE
JP - Overall User Progress and User Progress by Library

### PR DESCRIPTION
**Pull request by juanpedromoreno**
This PR brings a couple of new services to fetch the user progress.

Please, @rafaparadela @raulraja @dialelo @andyscott Review when you have a chance. 

Thanks.

### CODE REVIEW

File: site/shared/src/main/scala/shared/User.scala [hunk 1]
```diff
@@ -26,9 +26,8 @@ case class UserProgress(
 
 case class OverallUserProgressItem(
   libraryName: String,
-  sections:    Int,
-  completed:   Int,
-  color:       String
+  sections:    Long,
+  completed:   Boolean
```
Comment (dialelo) [inline site/shared/src/main/scala/shared/User.scala]: Note that `completed` cannot be inferred from the progress alone, we need to know if we have completed as many sections as they are available and the list of sections is only known at runtime.

Comment (juanpedromoreno) [inline site/shared/src/main/scala/shared/User.scala]: You are right, I missed that. I'll change it.

**Comment by juanpedromoreno** [comment]
@dialelo Take another look to the latest commit in order to double check now we are checking the completed field properly. Thanks.

**Comment by dialelo** [comment]
Looks right. 

For the client UI to be able to know about the state of every exercise it'll need that `fetchUserProgressByLibrarySection` includes the progress for every exercise in the section. Since unevaluated exercises are not stored we'll have to combine data from the exercises service and the user progress. What do you think @juanpedromoreno?

**Comment by juanpedromoreno** [comment]
@dialelo Let's take that information from the DOM as we talked ;)

**Comment by dialelo** [comment]
Looks good to me then, will merge as soon as Travis finishes!

**Comment by juanpedromoreno** [comment]
@dialelo Ready for another look. Now the endpoint returns all the exercises, even those exercises not completed by the user. Let me know if looks fine.

File: site/server/app/com/fortysevendeg/exercises/persistence/domain/UserProgressQueries.scala [hunk 2]
```diff
@@ -54,12 +54,28 @@ object UserProgressQueries {
 
   val findByUserId = s"$commonFindBy WHERE userId = ?"
 
-  val findByExerciseVersion =
-    s"""$commonFindBy WHERE userId = ? AND libraryName = ? AND sectionName = ? AND method = ? AND version = ?"""
+  val findByUserIdAggregated =
+    s"""
+       SELECT libraryname, count(sectionname), bool_and(succeeded)
```
Comment (raulraja) [inline site/server/app/com/fortysevendeg/exercises/persistence/domain/UserProgressQueries.scala]: Cool shit!

File: site/server/app/com/fortysevendeg/exercises/persistence/repositories/UserProgressRepository.scala [hunk 3]
```diff
@@ -53,14 +64,14 @@ class UserProgressDoobieRepository(implicit persistence: PersistenceModule) exte
   override def findByUserId(userId: Long): ConnectionIO[List[UserProgress]] =
     persistence.fetchList[Long, UserProgress](Q.findByUserId, userId)
 
-  override def findByExerciseVersion(
+  override def findByUserIdAggregated(userId: Long): ConnectionIO[List[(String, Long, Boolean)]] =
+    persistence.fetchList[Long, (String, Long, Boolean)](Q.findByUserIdAggregated, userId)
```
Comment (raulraja) [inline site/server/app/com/fortysevendeg/exercises/persistence/repositories/UserProgressRepository.scala]: We should create type aliases perhaps for these pairs of input and tupled replacements if these are in use elsewhere.

File: site/server/app/com/fortysevendeg/exercises/services/free/userprogressops.scala [hunk 4]
```diff
@@ -45,8 +69,39 @@ class UserProgressOps[F[_]](implicit I: Inject[UserProgressOp, F], EO: ExerciseO
     import ConnectionIOOps._
     for {
       lbs ← UPR.findUserProgressBySection(user, libraryName, sectionName).liftF[F]
-      sectionCount ← EO.getLibrary(libraryName) map (_ map (_.sections.size) getOrElse 0)
-    } yield LibrarySectionArgs(libraryName, sectionCount, lbs.exerciseList, lbs.succeeded)
+      libraryInformation ← EO.getLibrary(libraryName) map extractLibraryInformation(sectionName)
+      eList = calculateExerciseList(lbs.exerciseList, libraryInformation._2)
```
Comment (raulraja) [inline site/server/app/com/fortysevendeg/exercises/services/free/userprogressops.scala]: let's expand this nasty tuple so we can see what it's all about in the inside.

File: site/server/test/UserProgressQueriesSpec.scala [hunk 5]
```diff
@@ -19,6 +19,8 @@ class UserProgressQueriesSpec
   check(Query[Unit, UserProgress](Q.all))
   check(Query[Long, UserProgress](Q.findById))
   check(Query[Long, UserProgress](Q.findByUserId))
+  check(Query[Long, (String, Long, Boolean)](Q.findByUserIdAggregated))
```
Comment (raulraja) [inline site/server/test/UserProgressQueriesSpec.scala]: That's what I'm talking about the type aliases!

File: site/server/test/UserProgressRepositorySpec.scala [hunk 6]
```diff
@@ -60,6 +60,36 @@ class UserProgressRepositorySpec
...
+      repository.create(pair.request).transact(transactor).run
+
+      val upByLibrary =
+        repository.findByLibrary(
+          userId = pair.user.id,
+          libraryName = pair.request.libraryName
+        ).transact(transactor).run
+
+      upByLibrary.size shouldBe 1
+      upByLibrary.head._2 shouldBe pair.request.succeeded
```
Comment (raulraja) [inline site/server/test/UserProgressRepositorySpec.scala]: let's unpack these tuples

**Comment by raulraja** [comment]
LGTM once my comments are addressed

**Pull request merged**